design/baremetal-operator/bulk-set-bios-config.md

```diff
@@ -0,0 +1,228 @@
+`HostFirwareSettingsSpec` if that field is empty. Settings that have the
+`read-only` flag set will not be copied as these settings cannot be written
+to the BMC. Likewise, settings that have the `unique` flag set will also not
+be copied, these are settings like serial numbers or product ids for example,
+that should not be written to other servers.
+
+A user can update `writeableFirmwareSettings` to the desired values for the
+settings.  Likewise, `writeableFirmwareSettings` can be copied to multiple hosts
+of the same type to effectively clone the settings from a reference host and
+ensuring that all similar systems have the same BIOS configuration,
+i.e. "bulk set".
+
+The baremetal-operator will detect when `writeableFirmwareSettings` changes by
+comparing the name/value pairs to `currentFirmwareSettings`. When a change is
+detected, the BMO will add the new values to the Ironic clean-steps API. Before
+adding it though, the BMO will do validation checks on the new values using the
+Registry information in `currentFirmwareSettings`. For example, integer types
+will checked against minimum and maximum parameters and enumeration types will
+be compared against allowable values. The validation webhook can be leveraged
+here to log a message regarding the failure; the invalid BIOS Setting will not
+be added to clean-steps.
+
+When settings are added to the Ironic clean-steps API Ironic, Ironic will set
+the BIOS Configuration in the BMC as part of the manual cleaning process which
+requires a reboot. Manual clean occurs in the Preparing state of the
+BaremetalHost, the Host will re-enter this state from Ready/Available state
+whenever its config differs from the last one it stored.
+
+After applying the new BIOS Configuration, Ironic will then read and cache the
+new BIOS Settings, which can be retrieved by the BMO and used to update
+`currentFirmwareSettings`.
+
+### Implementation Details/Notes/Constraints
+
+One of the primary drivers of his design was that each vendor uses different
+names and values for similar BIOS Configuration settings. This design does not
+rely on vendor specific checks but instead uses common mechanisms to ensure
+the BIOS Settings are valid:
+
+- copying the names and values from `currentFirmwareSettings` to
+  `writeableFirmwareSettings` ensures initial values are correct
+- when updating the values in `writeableFirmwareSettings`, the user can use the
+  limitations in the `currentFirmwareSettings`, e.g. minimum and maximum values
+- before writing to the Ironic clean-steps API, the BMO runs validation checks
+  against the `currentFirmwareSettings` to ensure updated values are valid
+
+In order to interoperate with the individual BIOS Configuration set
+approach the BMO must ensure that the same settings don't get added twice
+to the Ironic clean-steps API. In addition, as each clean step requires a
+reboot the number of clean steps that are added must be managed.
+
+### Risks and Mitigations
+
+N/A
```

Comment:
I remember at some point we said there was no sensitive data being exposed, but I can't remember if we were proposing to show the read-only setting values at that time. Can you confirm that this is still the case?

Reply:
I think we should show the readOnly data as its not sensitive, its normally things like NumCores or CoreSpeed which can't be changed. However some vendors, Dell in particular, make use of the "Password" AttributeType defined in https://redfish.dmtf.org/schemas/v1/AttributeRegistry.v1_3_5.json and include settings like "SysPassword" and "SetupPassword" with this AttributeType. In this case it looks like the values returned are either not included or set to the empty string but we should not include settings of this type in the currentValues and disallow any sets of this AttributeType.

I will add a note to this section